### PR TITLE
I7218-Disco taar update

### LINKED ISSUE #7218: TAARDIS: Pass hashed client id to TAAR via AMO API request
As TAAR moves away from being a shield study instead of receiving the client-id via a query string set by the add-on the client-id will be available in hashed form as a cookie sent by the browser.

Bug with the details of the Firefox implementation is here: https://bugzilla.mozilla.org/show_bug.cgi?id=1489531

This bug is to cover reading that cookie and passing the hashed id down to TAAR via the AMO api.

### PR DESCRIPTION
**Pull request by rebmullin**
fixes #7218 

depends on ~~#7371 ('ll update this disco reducer once this gets merged)~~

### CODE REVIEW

File: config/default.js [hunk 1]
```diff
@@ -350,6 +352,10 @@ module.exports = {
 
   // Feature flags.
   //
+
+  enableFeatureDiscoTaar: false,
+  discoTaarIdCookie: 'taarId',
```
Comment (willdurand) [inline config/default.js]: Why is this located in the global config if it only applies to the `disco` app?

File: config/default.js [hunk 2]
```diff
@@ -350,6 +352,10 @@ module.exports = {
 
   // Feature flags.
   //
+
+  enableFeatureDiscoTaar: false,
```
Comment (willdurand) [inline config/default.js]: I haven't seen any mention of a feature flag for that. Is there an issue for that?

File: config/dev-disco.js [hunk 3]
```diff
@@ -25,4 +25,13 @@ module.exports = {
 
   // https://sentry.prod.mozaws.net/operations/addons-frontend-disco-dev/
   publicSentryDsn: 'https://560fc81d9fd14266b99bda032de23c52@sentry.prod.mozaws.net/184',
+
+  enableFeatureDiscoTaar: true,
+  discoParamsToUse: [
```
Comment (willdurand) [inline config/dev-disco.js]: Where does this come from? And why is it in the config?

File: config/development-disco.js [hunk 4]
```diff
@@ -1,4 +1,13 @@
 module.exports = {
   trackingEnabled: false,
   loggingLevel: 'debug',
+
+  enableFeatureDiscoTaar: true,
+  discoParamsToUse: [
```
Comment (willdurand) [inline config/development-disco.js]: Same question.

File: src/core/server/base.js [hunk 5]
```diff
@@ -42,6 +42,7 @@ import {
   langToLocale,
   makeI18n,
 } from 'core/i18n/utils';
+import { setTaarId } from 'disco/reducers/discoResults';
```
Comment (willdurand) [inline src/core/server/base.js]: Do we really want to set a TAAR `id` in the `disco` results? I'd create a new reducer for that. Also, we've used `clientId` (or client ID or telemetry client ID) for a while, so I would not change that name, because it is a bit confusing. Sure the cookie name in FF is `taarId` but that's just the cookie name. We can have `clientId` on our side.

I would create a `telemetry` reducer maybe and store the `clientId` in it. Because the content is hashed, maybe `hashedClientId` would be a better name and we would have an action creator named `setHashedClientId` for example.

File: src/disco/api.js [hunk 6]
```diff
@@ -12,10 +12,10 @@ export function getDiscoveryAddons({ api, taarParams = {}, _config = config }) {
     return object;
   }, {});
 
-  // We translate `clientId` to `'telemetry-client-id'`.
-  if (allowedTaarParams.clientId) {
-    allowedTaarParams['telemetry-client-id'] = allowedTaarParams.clientId;
-    delete allowedTaarParams.clientId;
+  // We translate `taarId` to `'telemetry-client-id'`.
```
Comment (willdurand) [inline src/disco/api.js]: As per my suggestion above, I would not change that. That would reduce the diff in this file.

File: src/core/server/base.js (hunk 5 above)
Comment (willdurand) [reply src/core/server/base.js]: That could be done in a PR before this one so that we would have one issue explaining why/how we plan to fully tackle #7218 at least (because understanding the aim of this PR was tough 😞)

Comment (rebmullin) [reply src/core/server/base.js]: > Do we really want to set a TAAR id in the disco results? I'd create a new reducer for that. Also, we've used clientId (or client ID or telemetry client ID) for a while, so I would not change that name, because it is a bit confusing. Sure the cookie name in FF is taarId but that's just the cookie name. We can have clientId on our side.

I thought about this (not using in the disco reducer) but I ultimately decided ti leave it here. I can definitely move it and do this in another PR like you suggest.

> That could be done in a PR before this one so that we would have one issue explaining why/how we plan to fully tackle #7218 at least (because understanding the aim of this PR was tough 😞)

Sorry, I should have put in a description above 🙈.

File: config/dev-disco.js (hunk 3 above)
Comment (rebmullin) [reply config/dev-disco.js]: I am overriding what's in `default-disco`. I was removing clientId but it's moot now so you said I should keep this as the same name. So I'll revert this: )

File: config/default.js (hunk 2 above)
Comment (rebmullin) [reply config/default.js]: @muffinresearch mentioned to me to set this up with a flag. 

@muffinresearch should we still do this?

File: config/default.js (hunk 1 above)
Comment (rebmullin) [reply config/default.js]: Do you mean both vars or just the cookie? Initially I had this in just disco (default-disco) but I think I got errors when I run amo:dev. I'll recheck this tho...

**Comment by rebmullin** [comment]
> Should I use FF Nightly to get the cookie and make sure everything works as intended?

I think you can just manually set a taarId cookie (try test id: 12b9377cbe7e5c94e8a70d9d23929523d14afa954793130f8a3959c7b849aca8) and you'll see it's passed to the API 

OR yes you should be able to see this in Nightly. I think the only thing you need to do is set the following for this:

```bash
+pref("browser.discovery.enabled", false); // note: if this doesn't work try true
+pref("browser.discovery.containers.enabled", true);
+pref("browser.discovery.sites", "example.com");
```
ref( https://hg.mozilla.org/mozilla-central/rev/374a4bf1482e)

but I don't *think* it's returning anything different yet (at least the test ids still don't seem to be working on dev at this point). See:

https://github.com/mozilla/addons-server/issues/10174#issuecomment-453469860

**Comment by willdurand** [comment]
This is almost perfect, thanks. There is room for improvement in the test code, but everything else looks 👌

File: src/core/server/base.js [hunk 7]
```diff
@@ -336,6 +337,18 @@ function baseServer(
           store.dispatch(dismissSurvey());
         }
 
+        if (
+          config.get('enableFeatureDiscoTaar') &&
+          req.universalCookies.get(config.get('discoTaarIdCookie')) !==
+            undefined
```
[the remark this answers is not in the document]
Comment (rebmullin) [reply src/core/server/base.js]: I thought about that but I didn't think I saw a config for that. I'll recheck this tho..

File: tests/unit/disco/test_api.js [hunk 8]
```diff
@@ -32,12 +32,13 @@ describe(__filename, () => {
       });
     });
 
-    it('calls the API with a telemetry client ID', () => {
-      const telemetryClientId = 'client-id';
+    it("passes through taarId as telemetry-client-id to the API when it's available", () => {
```
[the remark this answers is not in the document]
Comment (rebmullin) [reply tests/unit/disco/test_api.js]: I thought this was clearer..

File: src/core/server/base.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/server/base.js]: You can probably use `appName`, either in `config` or as an argument passed to the function that contains this logic.

File: tests/unit/disco/test_api.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply tests/unit/disco/test_api.js]: It's more a general question: no API code has been changed, so I don't expect any change in the test file of the API code.

File: tests/unit/disco/sagas/test_disco.js [hunk 9]
```diff
@@ -126,7 +138,7 @@ describe(__filename, () => {
       _getDiscoResults({
         taarParams: {
           platform: 'Darwin',
-          'telemetry-client-id': telemetryClientId,
+          clientId: taarId,
```
[the remark this answers is not in the document]
Comment (rebmullin) [reply tests/unit/disco/sagas/test_disco.js]: hmm..yeh..

File: tests/unit/core/server/test_base.js [hunk 10]
```diff
@@ -222,6 +225,106 @@ describe(__filename, () => {
       expect(survey.wasDismissed).toEqual(true);
     });
 
+    it('dispatches setHashedClientId() if cookie is present and enableFeatureDiscoTaar is true', async () => {
```
Comment (willdurand) [inline tests/unit/core/server/test_base.js]: ... and app is "disco"

File: tests/unit/core/server/test_base.js [hunk 11]
```diff
@@ -222,6 +225,106 @@ describe(__filename, () => {
...
+        sagaMiddleware,
+      })
+        .get('/en-US/firefox/')
+        .set('cookie', `${fakeConfig.get('discoTaarIdCookie')}="${clientId}"`)
+        .end();
+
+      sinon.assert.neverCalledWith(dispatchSpy, setHashedClientId(clientId));
+
+      expect(response.statusCode).toEqual(200);
+      expect(store.getState().telemetry.hashedClientId).toEqual(null);
```
Comment (willdurand) [inline tests/unit/core/server/test_base.js]: Is this possible? We should not have `telemetry` in the store when app is `disco`.

File: tests/unit/core/server/test_base.js [hunk 12]
```diff
@@ -222,6 +225,106 @@ describe(__filename, () => {
...
+    });
```
Comment (willdurand) [inline tests/unit/core/server/test_base.js]: You might want to put all these newly added test cases into a `describe` block:

```
describe('TAARDIS', () => {});
```

File: tests/unit/disco/pages/TestDiscoPane.js [hunk 13]
```diff
@@ -105,23 +106,21 @@ describe(__filename, () => {
       );
     });
 
-    it('sends a telemetry client ID if there is one', () => {
-      const location = createFakeLocation({
-        query: {
-          clientId: 'telemetry-client-id',
-        },
-      });
+    it('sends a telemetry-client-id if there is one', () => {
```
Comment (willdurand) [inline tests/unit/disco/pages/TestDiscoPane.js]: I liked `telemetry client ID` better.

File: tests/unit/core/server/test_base.js (hunk 11 above)
Comment (rebmullin) [reply tests/unit/core/server/test_base.js]: yeh good catch. I think I need to adjust the logic a bit.

File: tests/unit/disco/pages/TestDiscoPane.js (hunk 13 above)
Comment (rebmullin) [reply tests/unit/disco/pages/TestDiscoPane.js]: okay...

File: tests/unit/core/server/test_base.js (hunk 11 above)
Comment (rebmullin) [reply tests/unit/core/server/test_base.js]: telemetry *should* only be in the store when it IS disco..

**Comment by willdurand** [comment]
> OR yes you should be able to see this in Nightly. I think the only thing you need to do is set the following for this:
> 
> ```shell
> +pref("browser.discovery.enabled", false); // note: if this doesn't work try true
> +pref("browser.discovery.containers.enabled", true);
> +pref("browser.discovery.sites", "example.com");
> ```

so we should set the first pref to `true` in FF Nightly. The `taarId` cookie gets generated with "Session" in "Expires on" date.

The disco app is unable to read this cookie though, for some reason the `setHashedClientId()` action is not dispatched. If I create a similar cookie but with a date instead of session, the cookie is read.

**Comment by rebmullin** [comment]
> ok, I will retry then.

thank you!

**Comment by willdurand** [comment]
I can't make it work, the cookie is never read by the app for some reasons, using `yarn disco:https`.

**Comment by rebmullin** [comment]
now that we are do a check for disco I think I can move all the config updates for this to default disco.

**Comment by rebmullin** [comment]
> I can't make it work, the cookie is never read by the app for some reasons, using `yarn disco:https`

hmm..bummer. I am not sure why I see this being passed in and why changing the expiry date on your end would matter..?

just double checking that you have something similar to this url:
https://example.com:3000/en-US/firefox/discovery/pane/65.0/Darwin/normal ?

**Comment by willdurand** [comment]
@rebmullin could you maybe ask someone else to try this patch in FF Nightly? I don't really know why the cookie isn't read locally.

**Comment by rebmullin** [comment]
hey @willdurand, so @bobsilverberg set this up and he is seeing the cookie being passed too. We both verified using a `console.log(apiURL)` in `core/api/index.js` since we didn't see it in the terminal (like we do with amo). 

I am just curious, is this how you checked too?

**Comment by rebmullin** [comment]
@willdurand, I was using `example.com:3000/en-US/firefox/discovery/pane/65.0/Darwin/normal` locally to test this so I didn't notice that this would cause issues with `about:addons`. Since the cookie is samesite=lax, this poses a problem it seems. I can only see this when I disable this pref setting (`network.cookie.same-site.enabled`). 

I'll guess I'll have to look into this more :/

**Comment by willdurand** [comment]
So this PR is blocked by the implementation of the communication channel between the disco app and FF.

**Comment by willdurand** [comment]
I tried the new patch with the header-based approach. I updated this PR with this diff and everything works!

```diff
diff --git a/src/core/server/base.js b/src/core/server/base.js
index 0ef5c399b..b1e628ddd 100644
--- a/src/core/server/base.js
+++ b/src/core/server/base.js
@@ -337,17 +337,14 @@ function baseServer(
           store.dispatch(dismissSurvey());
         }
 
+        const mozClientIdHeader = 'moz-client-id';
+
         if (
           appName === 'disco' &&
           config.get('enableFeatureDiscoTaar') &&
-          req.universalCookies.get(config.get('discoTaarIdCookie')) !==
-            undefined
+          req.headers[mozClientIdHeader]
         ) {
-          store.dispatch(
-            setHashedClientId(
-              req.universalCookies.get(config.get('discoTaarIdCookie')),
-            ),
-          );
+          store.dispatch(setHashedClientId(req.headers[mozClientIdHeader]));
         }
 
         pageProps = getPageProps({ store, req, res, config });
```

**Comment by willdurand** [comment]
> I tried the [new patch with the header-based approach](https://phabricator.services.mozilla.com/D16601).

Instructions to build FF with this patch:

1. if you don't have `mozilla-central` locally yet, get it: https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Build_Instructions#Build_prerequisites. You can run `./mach bootstrap` but I don't think you need to run `./mach build` yet 
2. download the diff file: [`D16601.diff`](https://phabricator.services.mozilla.com/D16601?download=true)
3. Apply the diff with the following command using `hg` (mercurial) in `/path/to/mozilla-central`:

    ```
    hg import /path/to/D16601.diff --no-commit
    ```

4. Create a `mozlog` file in the `mozilla-central` repo:

    ```
    touch mozlog
    ```

5. Put the following content into this `mozlog` file to use [artifact builds](https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Build_Instructions/Artifact_builds):

    ```
    # Automatically download and use compiled C++ components:
    ac_add_options --enable-artifact-builds

    # Write build artifacts to:
    mk_add_options MOZ_OBJDIR=./objdir-frontend
    ```

6. Build FF:

    ```
    ./mach build
    ```

7. Run FF:

    ```
    ./mach run
    ```

File: src/core/server/base.js [hunk 14]
```diff
@@ -336,6 +340,16 @@ function baseServer(
           store.dispatch(dismissSurvey());
         }
 
+        const mozClientIdHeader = DISCO_TAAR_CLIENT_ID_HEADER;
```
Comment (willdurand) [inline src/core/server/base.js]: I suppose you can use the constant directly now, no need to create a variable here.

File: tests/unit/core/server/test_base.js [hunk 15]
```diff
@@ -451,6 +456,113 @@ describe(__filename, () => {
...
+        sagaMiddleware,
+      })
+        .get('/en-US/firefox/')
+        .end();
+
+      sinon.assert.neverCalledWith(dispatchSpy, setHashedClientId('1112'));
+
+      expect(response.statusCode).toEqual(200);
+      expect(store.getState().telemetry).toEqual(undefined);
+    });
```
Comment (willdurand) [inline tests/unit/core/server/test_base.js]: We could maybe duplicate this test case for `moz-client-id` set to `null` and `undefined` so that those edge cases are covered too.

Comment (rebmullin) [reply tests/unit/core/server/test_base.js]: I've added a case for null but I couldn't set moz-client-id to undefined

Comment (willdurand) [reply tests/unit/core/server/test_base.js]: ok that should be good enough

**Comment by rebmullin** [comment]
@willdurand, is it okay to remove the blocked label? Or should we wait to their PR is merged.

**Comment by rebmullin** [comment]
hmm it looks like they made some changes already so maybe we should wait some more..?

**Comment by rebmullin** [comment]
@willdurand, thanks for your help with this 1! ⭐️
